### PR TITLE
Support Puppet 7

### PR DESCRIPTION
**Pull request by ekohl**
https://puppet.com/docs/puppet/7.4/server/release_notes.html

* [x] Handle cadir move to `/etc/puppetlabs/puppetserver/ca`
* [x] Handle auth.conf removal
* [x] Handle JSON fact cache instead of YAML (as used by Foreman's ENC)
* [x] Look at any legacy Puppet 3 endpoints
* [x] Align default ciphers

### CODE REVIEW

**Comment by mattock** [comment]
@ekohl I'm working on this as well. I've issued a few rather trivial metadata.json PRs which make Kafo happy about Puppet version requirements.

I was able to run our own puppetmaster-installer fine, but it fails with authorization errors. I'll look into that more tomorrow.

If you have anything work in progress / unpushed let me know so that we can coordinate the work :smile:.

**Comment by ekohl** [comment]
I haven't had time so all my work is here. Fact uploading is certainly broken right now. I do have some code for it and started https://github.com/theforeman/puppet-foreman/pull/925 but puppetserver isn't writing out the files on my server and I couldn't figure it out until now.

**Comment by ekohl** [comment]
Also linking it back here: https://community.theforeman.org/t/puppet-7-support/22786

File: manifests/server/puppetserver.pp [hunk 1]
```diff
@@ -246,10 +246,16 @@
     content => template('puppet/server/puppetserver/conf.d/puppetserver.conf.erb'),
   }
 
-  $auth_template = pick($server_auth_template, 'puppet/server/puppetserver/conf.d/auth.conf.erb')
+  if versioncmp($server_puppetserver_version, '7.0.0') >= 0 {
+    $auth_ensure = 'absent'
```
Comment (mattock) [inline manifests/server/puppetserver.pp:250]: Ensuring absent on the puppetserver auth.conf breaks puppetserver based on my testing - it really needs an auth.conf to work. The default puppetserver auth.conf that is bundled with this Puppet module seems to work with Puppetserver 7 as well.

Any particular reason to ensure absence here?

Comment (ekohl) [reply manifests/server/puppetserver.pp:250]: No, this is very much a draft PR and I assumed from reading the changelog that it should be absent.

Comment (mattock) [reply manifests/server/puppetserver.pp:250]: Ok. Afaics this part can remain as it was. Only the legacy auth won't work, but that's being checked above and if it is being used catalog compilation will fail.

Comment (ekohl) [reply manifests/server/puppetserver.pp:250]: My current focus is the transition to https://github.com/theforeman/puppet-puppetserver_foreman for which most has been merged now. Then I'll try to get the Puppet 7 support there. If you have patches in the mean time, I have no problem if you take over this effort or provide patches in some way. Either PRs to my repo or directly, it doesn't matter.

Comment (mattock) [reply manifests/server/puppetserver.pp:250]: This module works well enough for our [use-case](https://forge.puppet.com/modules/puppetfinland/puppetmaster) so I probably won't be needing any fixes in the near future.

One thing that could potentially help you: you may need to adjust the cipher-list parameters for PuppetDB and Puppetserver. We had to do it to support Puppetboard. See [ssl_workarounds.pp](https://github.com/Puppet-Finland/puppet-puppetmaster/blob/master/manifests/puppetboard/ssl_workarounds.pp)  for details.

Comment (ekohl) [reply manifests/server/puppetserver.pp]: Fixed now. I deleted the wrong auth.conf. This one was the server but they meant the one that used to be shipped with the agent.

**Comment by ekohl** [comment]
I pushed the cadir handling, but haven't had a chance to test it. That's why it's still a draft and not checked off.

File: manifests/server/config.pp [hunk 2]
```diff
@@ -174,6 +174,18 @@
...
     }
+
+    # In Puppet 7 the cadir was changed from $ssldir/ca to $puppetserver_dir/ca
+    # This migrates the directory if it was in the old location
+    # The migration command leaves a symlink in place
+    if versioncmp($puppet::server::real_puppetserver_version, '7.0') > 0 {
+      exec { 'migrate Puppetserver cadir':
+        command => "${puppet::puppetserver_cmd} ca migrate",
+        creates => $puppet::server::cadir,
+        onlyif  => "test -d '${puppet::server::ssldir}/ca' && ! test -L '${puppet::server::ssldir}'",
```
Comment (ehelms) [inline manifests/server/config.pp:185]:
```suggestion
        onlyif  => "test -d '${puppet::server::ssl_dir}/ca' && ! test -L '${puppet::server::ssl_dir}'",
```

File: manifests/server/config.pp [hunk 3]
```diff
@@ -174,6 +174,18 @@
         Exec['puppet_server_config-create_ssl_dir'],
       ],
     }
+
+    # In Puppet 7 the cadir was changed from $ssldir/ca to $puppetserver_dir/ca
+    # This migrates the directory if it was in the old location
+    # The migration command leaves a symlink in place
+    if versioncmp($puppet::server::real_puppetserver_version, '7.0') > 0 {
+      exec { 'migrate Puppetserver cadir':
+        command => "${puppet::puppetserver_cmd} ca migrate",
```
Comment (ekohl) [inline manifests/server/config.pp:183]: This returns:
```
Notice: /Stage[main]/Puppet::Server::Config/Exec[migrate Puppetserver cadir]/returns: Puppetserver service is running. Please stop it before attempting to run this command.
Error: '/opt/puppetlabs/bin/puppetserver ca migrate' returned 1 instead of one of [0]
Error: /Stage[main]/Puppet::Server::Config/Exec[migrate Puppetserver cadir]/returns: change from 'notrun' to ['0'] failed: '/opt/puppetlabs/bin/puppetserver ca migrate' returned 1 instead of one of [0]
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: The cadir is currently configured to be inside the /etc/puppetlabs/puppet/ssl directory. This config setting and the directory location will not be used in a future version of puppet. Please run the puppetserver ca tool to migrate out from the puppet confdir to the /etc/puppetlabs/puppetserver/ca directory. Use `puppetserver ca migrate --help` for more info.
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Error:
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_crt.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_crl.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/infra_crl.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/certs/centos8.wisse.example.com.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/crl.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_pub.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/inventory.txt'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/infra_inventory.txt'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/infra_serials'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/serial'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/signed/centos8.wisse.example.com.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/root_key.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_key.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: If you would really like to replace your CA, please delete the existing files first.
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Note that any certificates that were issued by this CA will become invalid if you
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: replace it!
Error: '/opt/puppetlabs/bin/puppetserver ca setup' returned 1 instead of one of [0]
Error: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: change from 'notrun' to ['0'] failed: '/opt/puppetlabs/bin/puppetserver ca setup' returned 1 instead of one of [0]
Notice: /Stage[main]/Puppet::Server::Service/Service[puppetserver]: Dependency Exec[puppet_server_config-generate_ca_cert] has failures: true
Notice: /Stage[main]/Puppet::Server::Service/Service[puppetserver]: Dependency Exec[migrate Puppetserver cadir] has failures: true
```

Tomorrow I'll hopefully be able to come up with a solution.

Comment (ekohl) [reply manifests/server/config.pp:183]: I've now added an item to README. For the installer I'll write a hook.

**Comment by ekohl** [comment]
I believe this is now ready. Let's see if the tests agree.

File: spec/acceptance/puppetserver_upgrade_spec.rb [hunk 4]
```diff
@@ -13,29 +13,43 @@
...
+  case ENV['BEAKER_PUPPET_COLLECTION']
+  when 'puppet7'
+    from_version = '7.0.0'
+    to_version = '7.2.0'
+  when 'puppet6'
+    from_version = '6.7.0'
+    to_version = '6.7.2'
+  when 'puppet5'
+    from_version = '5.3.6'
+    to_version = '5.3.7'
```
Comment (ekohl) [inline spec/acceptance/puppetserver_upgrade_spec.rb:25]: Somehow this finds puppet6 on Debian builds but not on CentOS. I don't understand why yet. There is a default is set [here](https://github.com/voxpupuli/voxpupuli-acceptance/blob/f780900191c2f6385eda09cdec5383c112261e34/lib/voxpupuli/acceptance/spec_helper_acceptance.rb#L5) so perhaps it's not propagating somehow.

File: spec/acceptance/puppetserver_upgrade_spec.rb [hunk 5]
```diff
@@ -13,29 +13,43 @@
...
+  else
+    raise 'Unsupported Puppet collection'
+  end
+
   case fact('osfamily')
   when 'Debian'
-    from_version = "5.3.6-1#{fact('lsbdistcodename')}"
-    to_version = "5.3.7-1#{fact('lsbdistcodename')}"
+    from_version_exact = "6.7.0-1#{fact('lsbdistcodename')}"
+    to_version_exact = "6.7.2-1#{fact('lsbdistcodename')}"
```
Comment (ehelms) [inline spec/acceptance/puppetserver_upgrade_spec.rb:33]: @ekohl Is this not the reason it is finding puppet6 on Debian? As this code says, when its Debian always use these versions.

Comment (ekohl) [reply spec/acceptance/puppetserver_upgrade_spec.rb]: That explains it. I'm not sure how I missed that when writing it.

**Comment by ekohl** [comment]
I think the max open file limit is failing on Puppet 5 because it's pulling in camptocamp/systemd 3 (bd359f0025fd3db30b8b1b98e86bde47a7043965) which has dropped support for Puppet 5. However, it is odd that it's failing here and not in master.

**Comment by ekohl** [comment]
CentOS 6 build failures are expected, just like Puppet 5 + Ubuntu 20.04. Rest is still running. If they're green, I think this should be merged.

**Comment by ehelms** [comment]
Looks like down to just `CI / Acceptance / Puppet 7 - Ubuntu 20.04 (pull_request)`

**Comment by ekohl** [comment]
Looks like this is repeatable. The debug log:

```
  Notice: /Stage[main]/Puppet::Server::Puppetserver/Systemd::Dropin_file[puppetserver.service-limits.conf]/File[/etc/systemd/system/puppetserver.service.d]/ensure: created
  Debug: /Stage[main]/Puppet::Server::Puppetserver/Systemd::Dropin_file[puppetserver.service-limits.conf]/File[/etc/systemd/system/puppetserver.service.d]: The container Systemd::Dropin_file[puppetserver.service-limits.conf] will propagate my refresh event
  Notice: /Stage[main]/Puppet::Server::Puppetserver/Systemd::Dropin_file[puppetserver.service-limits.conf]/File[/etc/systemd/system/puppetserver.service.d/limits.conf]/ensure: defined content as '{sha256}af70de30e188923816fe1eb77708cdc603f53918edd1f7815bea5bc48690f062'
  Debug: /Stage[main]/Puppet::Server::Puppetserver/Systemd::Dropin_file[puppetserver.service-limits.conf]/File[/etc/systemd/system/puppetserver.service.d/limits.conf]: The container Systemd::Dropin_file[puppetserver.service-limits.conf] will propagate my refresh event
  Debug: Systemd::Dropin_file[puppetserver.service-limits.conf]: The container Class[Puppet::Server::Puppetserver] will propagate my refresh event
  Debug: Class[Puppet::Server::Puppetserver]: The container Stage[main] will propagate my refresh event
  Debug: Class[Puppet::Server::Puppetserver]: The container Class[Puppet::Server::Config] will propagate my refresh event
  Debug: Class[Puppet::Server::Config]: The container Stage[main] will propagate my refresh event
  Debug: Class[Puppet::Server::Config]: The container Class[Puppet::Server] will propagate my refresh event
  Info: Class[Puppet::Server::Config]: Scheduling refresh of Class[Puppet::Server::Service]
  Info: Class[Puppet::Server::Service]: Scheduling refresh of Service[puppetserver]
  Debug: Executing: '/usr/bin/systemctl is-active -- puppetserver'
  Debug: Executing: '/usr/bin/systemctl is-enabled -- puppetserver'
  Debug: Executing: '/usr/bin/systemctl is-active -- puppetserver'
  Debug: Executing: '/usr/bin/systemctl show --property=NeedDaemonReload -- puppetserver'
  Debug: Executing: '/usr/bin/systemctl restart -- puppetserver'
```

However, when I run it later:
```console
# systemctl cat puppetserver
# Warning: puppetserver.service changed on disk, the version systemd has loaded is outdated.
# This output shows the current version of the unit's original fragment and drop-in files.
# If fragments or drop-ins were added or removed, they are not properly reflected in this output.
# Run 'systemctl daemon-reload' to reload units.
# /lib/systemd/system/puppetserver.service
...
# systemctl show --property=NeedDaemonReload -- puppetserver
NeedDaemonReload=yes
# cat /etc/systemd/system/puppetserver.service.d/limits.conf
[Service]
LimitNOFILE=32143
```

This looks like a real edge case and I'm not sure where it comes from. Race condition with systemd itself?

**Comment by ehelms** [comment]
> This looks like a real edge case and I'm not sure where it comes from. Race condition with systemd itself?

Fun. I had seen this similar error with a Puppet 5 test so I agree it must be a bit of a race condition.

**Review by ehelms — APPROVED** [review]
Will let you merge in case anything needs to be squashed.

**Comment by ekohl** [comment]
I found a minimal reproducer:
```console
# ls /etc/systemd/system/puppetserver.service.d
limits.conf
mv /etc/systemd/system/puppetserver.service.d ~ && systemctl show --property=NeedDaemonReload -- puppetserver
NeedDaemonReload=yes
# systemctl daemon-reload
# mv ~/puppetserver.service.d /etc/systemd/system/puppetserver.service.d && systemctl show --property=NeedDaemonReload -- puppetserver
NeedDaemonReload=no
# systemctl cat puppetserver
# /lib/systemd/system/puppetserver.service
#
# Local settings can be configured without being overwritten by package upgrades, for example
# if you want to increase puppetserver open-files-limit to 10000,
# you need to increase systemd's LimitNOFILE setting, so create a file named
# "/etc/systemd/system/puppetserver.service.d/limits.conf" containing:
#       [Service]
#       LimitNOFILE=10000
# You can confirm it worked by running systemctl daemon-reload
# then running systemctl show puppetserver | grep LimitNOFILE
#
[Unit]
Description=puppetserver Service
After=syslog.target network.target nss-lookup.target 

[Service]
Type=forking
EnvironmentFile=/etc/default/puppetserver
User=puppet
TimeoutStartSec=300
TimeoutStopSec=60
Restart=on-failure
StartLimitBurst=5
PIDFile=/var/run/puppetlabs/puppetserver/puppetserver.pid

# https://tickets.puppetlabs.com/browse/EZ-129
# Prior to systemd v228, TasksMax was unset by default, and unlimited. Starting in 228 a default of '512'
# was implemented. This is low enough to cause problems for certain applications. In systemd 231, the
# default was changed to be 15% of the default kernel limit. This explicitly sets TasksMax to 4915,
# which should match the default in systemd 231 and later.
# See https://github.com/systemd/systemd/issues/3211#issuecomment-233676333
TasksMax=4915

#set default privileges to -rw-r-----
UMask=027


ExecReload=/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver reload
ExecStart=/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver start
ExecStop=/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver stop

KillMode=process

SuccessExitStatus=143

StandardOutput=syslog

[Install]
WantedBy=multi-user.target
# systemctl daemon-reload
# systemctl cat puppetserver
# /lib/systemd/system/puppetserver.service
#
# Local settings can be configured without being overwritten by package upgrades, for example
# if you want to increase puppetserver open-files-limit to 10000,
# you need to increase systemd's LimitNOFILE setting, so create a file named
# "/etc/systemd/system/puppetserver.service.d/limits.conf" containing:
#       [Service]
#       LimitNOFILE=10000
# You can confirm it worked by running systemctl daemon-reload
# then running systemctl show puppetserver | grep LimitNOFILE
#
[Unit]
Description=puppetserver Service
After=syslog.target network.target nss-lookup.target 

[Service]
Type=forking
EnvironmentFile=/etc/default/puppetserver
User=puppet
TimeoutStartSec=300
TimeoutStopSec=60
Restart=on-failure
StartLimitBurst=5
PIDFile=/var/run/puppetlabs/puppetserver/puppetserver.pid

# https://tickets.puppetlabs.com/browse/EZ-129
# Prior to systemd v228, TasksMax was unset by default, and unlimited. Starting in 228 a default of '512'
# was implemented. This is low enough to cause problems for certain applications. In systemd 231, the
# default was changed to be 15% of the default kernel limit. This explicitly sets TasksMax to 4915,
# which should match the default in systemd 231 and later.
# See https://github.com/systemd/systemd/issues/3211#issuecomment-233676333
TasksMax=4915

#set default privileges to -rw-r-----
UMask=027


ExecReload=/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver reload
ExecStart=/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver start
ExecStop=/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver stop

KillMode=process

SuccessExitStatus=143

StandardOutput=syslog

[Install]
WantedBy=multi-user.target

# /etc/systemd/system/puppetserver.service.d/limits.conf
[Service]
LimitNOFILE=32143
```

So it looks like the `NeedDaemonReload` value is not correct if you create a directory with a file in it. I suspect a create directory -> create override may be the same problem.

For now I'll add the daemon reload as a workaround, but this may be a systemd bug.
